### PR TITLE
Create generic repository and unit of work

### LINKED ISSUE #15: Create Generic Repositoy and Unit Of Work

### PR DESCRIPTION
**Pull request by hnthinh55**

### CODE REVIEW

File: TicketBooking.Data/Infrastructure/GenericRepository.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+using TicketBooking.Data.DbContext;
+
+namespace TicketBooking.Data.Infrastructure
+{
+    public class GenericRepository<T, X> : IRepository<T, X> where T : class where X : Type
+    {
```
Comment (tanngocIT) [inline TicketBooking.Data/Infrastructure/GenericRepository.cs:13]: rename to Repository

File: TicketBooking.Data/DbContext/TicketBookingDbContext.cs [hunk 2]
```diff
@@ -11,7 +11,15 @@ namespace TicketBooking.Data.DbContext
 {
     public class TicketBookingDbContext : IdentityDbContext<ApplicationUser>
     {
+        public static TicketBookingDbContext Create()
+        {
```
Comment (tanngocIT) [inline TicketBooking.Data/DbContext/TicketBookingDbContext.cs]: Why do we need this method?

File: TicketBookingAPI/Program.cs [hunk 3]
```diff
@@ -11,7 +12,7 @@
 builder.Services.AddSwaggerGen();
 builder.Services.AddDbContext<TicketBookingDbContext>(op =>
     op.UseSqlServer(builder.Configuration.GetConnectionString("DefaultConnection")));
-
+builder.Services.AddTransient<IUnitOfWork, UnitOfWork>();
```
Comment (tanngocIT) [inline TicketBookingAPI/Program.cs]: why don't we use .AddScoped?

File: TicketBooking.Data/DbContext/TicketBookingDbContext.cs (hunk 2 above)
Comment (hnthinh55) [reply TicketBooking.Data/DbContext/TicketBookingDbContext.cs]: That method is from a guide we follow a long time ago but now it seems not necessary so we already remove it

File: TicketBooking.Data/Infrastructure/IRepository.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TicketBooking.Data.Infrastructure
+{
+    public interface IRepository<T,X> where T : class where X : class
+    {
```
Comment (tanngocIT) [inline TicketBooking.Data/Infrastructure/IRepository.cs:11]: can we add more methods for included and order result?

File: TicketBookingAPI/Program.cs (hunk 3 above)
Comment (hnthinh55) [reply TicketBookingAPI/Program.cs]: Yes, this is our mistake. AddScoped will give better performance than addTransient, we did not notice this issue and will fix in next commit.

File: TicketBooking.Data/Infrastructure/IRepository.cs (hunk 4 above)
Comment (hnthinh55) [reply TicketBooking.Data/Infrastructure/IRepository.cs:11]: Sorry we do not get your question. You mean we should add methods that could get results in order?

Comment (tanngocIT) [reply TicketBooking.Data/Infrastructure/IRepository.cs:11]: Yes. the method support ordering the result. Load relationships model.

File: TicketBookingAPI/Program.cs (hunk 3 above)
Comment (tanngocIT) [reply TicketBookingAPI/Program.cs]: should change and complete in one PR

**Review by tanngocIT — APPROVED** [review]

**Pull request merged**